psycopg/psycopg/errors.py

```diff
@@ -18,19 +18,210 @@
 
 # Copyright (C) 2020 The Psycopg Team
 
-from typing import Any, Dict, Optional, Sequence, Tuple, Type, Union
+from dataclasses import dataclass, fields
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    List,
+    NoReturn,
+    Optional,
+    Sequence,
+    Tuple,
+    Type,
+    Union,
+    TYPE_CHECKING,
+)
 from typing_extensions import TypeAlias
 from asyncio import CancelledError
 
 from .pq.abc import PGconn, PGresult
 from .pq._enums import DiagnosticField
 from ._compat import TypeGuard
 
+if TYPE_CHECKING:
+    from .pq.misc import PGnotify, ConninfoOption
+
 ErrorInfo: TypeAlias = Union[None, PGresult, Dict[int, Optional[bytes]]]
 
 _sqlcodes: Dict[str, "Type[Error]"] = {}
 
 
+@dataclass
+class FinishedPGconn:
+    """Finished libpq connection.
+
+    Attributes are set from a real `~pscopg.pq.PGconn` but any operations will
+    raise an `~psycopg.OperationalError`.
+    """
+
+    info: List["ConninfoOption"]
+
+    db: bytes
+    user: bytes
+    password: bytes
+    host: bytes
+    port: bytes
+    tty: bytes
+    options: bytes
+    status: int
+    transaction_status: int
+
+    error_message: bytes
+    server_version: int
+
+    backend_pid: int
+    needs_password: bool
+    used_password: bool
+    ssl_in_use: bool
+
+    nonblocking: int
+
+    hostaddr: bytes = b""  # PostgreSQL >= 12
+    pipeline_status: int = 0  # PostgreSQL >= 14
+
+    notice_handler: Optional[Callable[["PGresult"], None]] = None
+    notify_handler: Optional[Callable[["PGnotify"], None]] = None
+
+    @classmethod
+    def connect(cls, *args: Any) -> NoReturn:
+        raise TypeError(f"{cls} is unusable")
+
+    @classmethod
+    def connect_start(cls, *args: Any) -> NoReturn:
+        raise TypeError(f"{cls} is unusable")
+
+    def connect_poll(self) -> NoReturn:
+        raise OperationalError("the connection is closed")
+
+    def finish(self) -> None:
+        pass
+
+    def reset(self) -> NoReturn:
+        raise OperationalError("the connection is closed")
+
+    def reset_start(self) -> NoReturn:
+        raise OperationalError("the connection is closed")
+
+    def reset_poll(self) -> NoReturn:
+        raise OperationalError("the connection is closed")
+
+    @classmethod
+    def ping(cls, *args: Any) -> NoReturn:
+        raise TypeError(f"{cls} is unusable")
```

Comment:
It's different because it applies to class methods, not to instance methods. I felt it would not make much sense to tell that the "connection is closed" when trying to open it (e.g. in `connect()`); and the `TypeError` feels more appropriate to me in that case (indicating that one cannot use this class for that purpose). We can change the message though.

Reply:
You are right. No problem with that.